Improve plot view

peci1 opened the pull request:
This PR improves handling of array and time types in plot_view.

This PR adds the following features:

1. It handles `array.array` and `numpy.ndarray` similar to `list` and `tuple`
2. It limits the maximum number of subobjects for list-like items to `MAX_LIST_LEN` + `LIST_TAIL_LEN`. Items on indices between `MAX_LIST_LEN` and `end - LIST_TAIL_LEN` are omitted.
3. If a list-like object has variable size (like `CompressedImage.data`), the user can select an index that is not present in all messages. In such case, the missing data are ignored and nothing is put on the timeline for that particular time.
4. It adds a virtual field `*float_seconds` to time message. This field contains the time value converted to float.

## Screenshots

### Current state
![Snímek obrazovky pořízený 2025-05-07 01-46-47](https://github.com/user-attachments/assets/714c32b9-6a7c-44f7-b7b1-40bde69f7c9c)

### New behavior

![Snímek obrazovky pořízený 2025-05-07 01-43-49](https://github.com/user-attachments/assets/964255d0-ff99-405a-8813-6de45a4b0ec5)
![Snímek obrazovky pořízený 2025-05-07 01-44-00](https://github.com/user-attachments/assets/0caa4482-744c-4e79-8a22-a82782fb6315)
![Snímek obrazovky pořízený 2025-05-07 01-45-00](https://github.com/user-attachments/assets/958a0980-7065-4f55-a302-43dc88eebd1a)

peci1 committed:
Improved handling of arrays in plot_view

Signed-off-by: Martin Pecka <peckama2@fel.cvut.cz>

peci1 committed:
Improved handling of time objects in plot_view

Signed-off-by: Martin Pecka <peckama2@fel.cvut.cz>

peci1 committed:
Fixed plot_view for items of variable-sized arrays.

Signed-off-by: Martin Pecka <peckama2@fel.cvut.cz>

rhaschke commented on rqt_bag_plugins/src/rqt_bag_plugins/plot_view.py:
```diff
@@ -432,18 +449,31 @@ def _add_msg_object(self, parent, path, name, obj, obj_type):
+                subobjs.append(('*float_seconds', time_obj.nanoseconds * 1e-9))
+        elif type(obj) in (list, tuple, array.array, numpy.ndarray):
+            if type(obj) in (array.array, numpy.ndarray):
+                list_obj = obj.tolist()
```
Explicitly converting the array into a list is a rather costly operation. You should operate directly on a reshaped array.

rhaschke commented on rqt_bag_plugins/src/rqt_bag_plugins/plot_view.py:
```diff
@@ -432,18 +449,31 @@ def _add_msg_object(self, parent, path, name, obj, obj_type):
                 w = int(math.ceil(math.log10(len_obj)))
-                subobjs = [('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(obj)]
+                subobjs = [('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(short_list_obj)]
+                if len_obj > MAX_LIST_LEN:
+                    for i in range(-LIST_TAIL_LEN, 0):
+                        if len_obj + i >= MAX_LIST_LEN:
+                            subobjs.append(('[%*d]' % (w, len_obj + i), list_obj[i]))
```
Can't you use the same logic here as before:
```suggestion
                tail_start = max(MAX_LIST_LEN, len_obj - LIST_TAIL_LEN)
                subobjs.extend([('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(list_obj[tail_start:])])
```

rhaschke reviewed (commented):
Generally looks good to me. I added a few remarks.
Great idea to make timestamp accessible as `float_seconds`. Maybe the name can be improved. What about just `time` or `seconds`?

peci1 committed:
plot_view: Better performance for arrays.

Signed-off-by: Martin Pecka <peckama2@fel.cvut.cz>

peci1 commented:
Thanks for the review, I've updated this PR.

> Maybe the name can be improved. What about just `time` or `seconds`?

I wanted to make it super-explicit that this is a virtual value. So it was intention to make such a name that cannot be mistakenly interpreted as a field name. But I've extracted the name into a constant, so it would be easier to change it if needed.

rhaschke commented on rqt_bag_plugins/src/rqt_bag_plugins/plot_view.py:
```diff
@@ -82,6 +88,10 @@
 
 from rqt_plot.data_plot import DataPlot
 
+MAX_LIST_LEN = 50
+LIST_TAIL_LEN = 10
```
Is it reasonable to make those variables ROS parameters?

peci1 replied:
In my view, that would be unnecessary complication.

rhaschke commented on rqt_bag_plugins/src/rqt_bag_plugins/plot_view.py:
```diff
@@ -432,20 +451,28 @@ def _add_msg_object(self, parent, path, name, obj, obj_type):
+                subobjs.append((FLOAT_SECONDS_LABEL, bag_helper.to_sec(time_obj)))
+        elif isinstance(obj, (Sequence, numpy.ndarray)) and not isinstance(obj, str):
             len_obj = len(obj)
+            short_obj = obj[:MAX_LIST_LEN]
+
```
We can directly access the head below (to use the same scheme as for tail):
```suggestion
```

rhaschke commented on rqt_bag_plugins/src/rqt_bag_plugins/plot_view.py:
```diff
@@ -432,20 +451,28 @@ def _add_msg_object(self, parent, path, name, obj, obj_type):
             else:
                 w = int(math.ceil(math.log10(len_obj)))
-                subobjs = [('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(obj)]
+                subobjs = [('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(short_obj)]
```
```suggestion
                subobjs = [('[%*d]' % (w, i), subobj) for (i, subobj) in enumerate(obj[:MAX_LIST_LEN])]
```

peci1 replied:
Okay, I went this way. God, I wish the line length limit were 120 and not 100 :)

rhaschke reviewed (approved):
Looks good to me. One small nitpick remaining.

peci1 committed:
Small refactoring.

Signed-off-by: Martin Pecka <peckama2@fel.cvut.cz>

peci1 commented:
Thanks, suggestion applied.

ahcorde commented:
https://github.com/Mergifyio backport kilted jazzy
